token_examples/factory_token/src/lib.rs

```diff
@@ -0,0 +1,87 @@
+use frc42_dispatch::match_method;
+use fvm_actor_utils::blockstore::Blockstore;
+use fvm_ipld_encoding::tuple::{Deserialize_tuple, Serialize_tuple};
+use fvm_sdk::NO_DATA_BLOCK_ID;
+use fvm_shared::{address::Address, error::ExitCode};
+pub mod token;
+
+use token::{deserialize_params, frc46_invoke, return_ipld, BasicToken, MintParams};
+
+#[derive(Serialize_tuple, Deserialize_tuple, Debug)]
+pub struct ConstructorParams {
+    pub name: String,
+    pub symbol: String,
+    pub granularity: u64,
+    /// authorised mint operator
+    /// only this address can mint tokens or remove themselves to permanently disable minting
+    pub minter: Option<Address>,
+}
+
+fn construct_token(params: ConstructorParams) {
+    let bs = Blockstore::default();
+    let token = BasicToken::new(&bs, params.name, params.symbol, params.granularity, params.minter);
+    let cid = token.save().unwrap();
+    fvm_sdk::sself::set_root(&cid).unwrap();
+}
+
+#[no_mangle]
+pub fn invoke(params: u32) -> u32 {
+    std::panic::set_hook(Box::new(|info| {
+        fvm_sdk::vm::abort(ExitCode::USR_ASSERTION_FAILED.value(), Some(&format!("{}", info)))
+    }));
+
+    let method_num = fvm_sdk::message::method_number();
+    // we only implement our own methods in this handler, anything we don't handle directly is delegated to frc46_invoke
+    // which handles any methods in the FRC46 token interface to save us writing the same
+    match_method!(method_num, {
+        "Constructor" => {
+            let params = deserialize_params(params);
+            construct_token(params);
+            NO_DATA_BLOCK_ID
+        }
+        "Mint" => {
+            let root_cid = fvm_sdk::sself::root().unwrap();
```

Comment:
I've changed it to have invoke() abort on error results and moved the method dispatching into a new function so it can return results. Clippy is mad about the size of the error type but that doesn't seem to be related to changes in this PR - I see it happening for https://github.com/helix-onchain/filecoin/pull/166 as well. I'll address that in a separate branch.